Adds illegal invocation error information (#15541)
cloudflare/cloudflare-docs: content/workers/observability/errors.md

}
```

### "Illegal invocation" errors

> The error message `TypeError: Illegal invocation: function called with incorrect \`this\` reference` can be a source of confusion.

Reviewer: ```suggestion
The error message `TypeError: Illegal invocation: function called with incorrect this reference` can be a source of confusion.
```

> This is typically caused by calling a member function of an object without the object itself as the receiver. For example, `let foo = obj.foo; foo();`. This is standard behavior in Javscript.

Reviewer: I think we can be clearer here and use words people actually know, rather than "member function of an object without the object itself as the receiver"

Reviewer: i.e. — when you destructure an object, like this:

```
let foo = obj.bar;
await foo();
```

...when you call `foo()`, the value of `this` is different than it would be if you called `obj.bar()`.

Author: Took another pass - LMK if you prefer it

}
```

### "Illegal invocation" errors

The error message `TypeError: Illegal invocation: function called with incorrect this reference` can be a source of confusion.

This is typically caused by calling a function that calls `this`, but the value of `this` has been lost.

For example, if `obj.foo()` is a method, and within that method `this` references `obj`, executing the method via `obj.foo();` will work properly. However, assigning it to a variable, `let func = obj.foo;`, then calling it, `func();`, will result in `this` being undefined. This is because `this` is lost when the method is called as a standalone function. This is standard behavior in JavaScript.

In practice, this is often seen when destructuring Javscript objects that have functions that rely on the presence of `this`, such as `ctx`. Avoid destructuring or re-bind the function to
the original context to avoid the error.

```js
---
filename: index.js
---
export default {
  async fetch(request, env, ctx) {

>     // destructuring ctx makes waitUntil lose its receiver

Reviewer 2: we never used the world `receiver` here, maybe we should keep this consistent with the above info?

```suggestion
    // destructuring ctx makes waitUntil lose its 'this' reference
```

    let { waitUntil } = ctx;
    // waitUntil errors, as it has no 'this'
    waitUntil(somePromise);

    // re-binding to ctx via apply, call, or bind avoids the error
    waitUntil.apply(ctx, [somePromise]);
    waitUntil.call(ctx, somePromise);

>     let reboundWaitUntil = waitUntil.bind(ctx)
>     reboundWaitUntil(someAsyncFunction)

Reviewer 2: ```suggestion
    const reboundWaitUntil = waitUntil.bind(ctx);
    reboundWaitUntil(somePromise);
```

>     // not destructuring ctx at all avoids the error
>     ctx.waitUntil(someAsyncFunction)

Reviewer 2: ```suggestion
    // directly calling the method on ctx avoids the error
    ctx.waitUntil(somePromise);
```
